### PR TITLE
Go: Implement connection logic in lib.rs

### PR DESCRIPTION
**Pull request by jonathanl-bq**

### CODE REVIEW

File: go/glide/lib.h [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  XAdd = 76,
+  XReadGroup = 77,
+  XAck = 78,
+  XTrim = 79,
+  XGroupCreate = 80,
+  XGroupDestroy = 81,
+};
+typedef uint32_t RequestType;
+
+typedef struct Level Level;
```
Comment (Yury-Fridlyand) [inline go/glide/lib.h:100]:
```suggestion
typedef struct _Level Level;
```
What is `_Level` BTW?

File: go/glide/lib.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+ * # Safety
+ * Unsafe function because creating string from pointer
+ */
+void log_ffi(struct Level log_level, const char *log_identifier, const char *message);
+
+/**
+ * # Safety
+ * Unsafe function because creating string from pointer
+ */
+struct Level init(struct Option_Level level, const char *file_name);
```
Comment (Yury-Fridlyand) [inline go/glide/lib.h:144]: why we need to return level?
```suggestion
void init(struct Option_Level level, const char *file_name);
```

File: go/src/lib.rs [hunk 3]
```diff
@@ -1,9 +1,434 @@
...
+                .expect("Can not read message argument."),
+        );
+    }
+}
+
+#[no_mangle]
+#[allow(improper_ctypes_definitions)]
+/// # Safety
+/// Unsafe function because creating string from pointer
+pub unsafe extern "C" fn init(level: Option<Level>, file_name: *const c_char) -> Level {
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:418]:
```suggestion
pub unsafe extern "C" fn init(level: Option<Level>, file_name: *const c_char) -> () {
```

File: go/examples/main.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	"fmt"
+	"github.com/aws/glide-for-redis/go/glide/glide"
+	"github.com/aws/glide-for-redis/go/glide/protobuf"
+)
+
+func main() {
+	fmt.Println("Starting go-glide client...")
+	client := glide.GlideRedisClient{}
+
+	request := &protobuf.ConnectionRequest{
```
Comment (acarbonetto) [inline go/examples/main.go:16]: Add a TODO for us to update this once Aaron's change is complete

File: go/examples/main.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		TlsMode:            protobuf.TlsMode_NoTls,
+		ClusterModeEnabled: false,
+		ReadFrom:           protobuf.ReadFrom_Primary,
+	}
+        request.Addresses = append(
+		request.Addresses,
+		&protobuf.NodeAddress{Host: "localhost", Port: uint32(6379)},
+	)
+	err := client.ConnectToRedis(request)
```
Comment (acarbonetto) [inline go/examples/main.go:25]: what happens if a user calls client.set() before they call ConnectToRedis?

File: go/examples/main.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"fmt"
+	"github.com/aws/glide-for-redis/go/glide/glide"
+	"github.com/aws/glide-for-redis/go/glide/protobuf"
+)
+
+func main() {
+	fmt.Println("Starting go-glide client...")
+	client := glide.GlideRedisClient{}
```
Comment (acarbonetto) [inline go/examples/main.go:14]: More consistent if we had a CreateClient() static method that called the constructor and returns:
```
client, error = glide.CreateClient(config);
```

Comment (acarbonetto) [reply go/examples/main.go:14]: then we can make ConnectToRedis internal

**Review by acarbonetto — COMMENTED** [review]
Let's merge this PR with Aaron's glide.go PR. 
Add IT tests for connection. 
Remove the set/get/ping/info content for a follow-up PR.

File: go/glide/lib.h (hunk 1 above)
Comment (jonathanl-bq) [reply go/glide/lib.h:100]: Logging level copied from your PR

File: go/glide/lib.h [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  XGroupCreate = 80,
+  XGroupDestroy = 81,
+};
+typedef uint32_t RequestType;
+
+typedef struct Level Level;
+
+typedef struct Option_Level Option_Level;
+
+typedef struct RedisErrorFFI {
```
[the remark this answers is not in the document]
Comment (jonathanl-bq) [reply go/glide/lib.h:104]: This shouldn't be a problem, since that's what cbindgen generates by default

File: go/glide/lib.h (hunk 2 above)
Comment (jonathanl-bq) [reply go/glide/lib.h:144]: idk I copied this from your PR

File: go/.gitignore [hunk 8]
```diff
@@ -1,4 +1,7 @@
 # Go compilation files
 *.pb.go
 
+# Examples
+examples/examples
```
[the remark this answers is not in the document]
Comment (jonathanl-bq) [reply go/.gitignore:5]: examples is the name of the binary generated when I run `go build` on the example. It probably just looked at the parent folder's name. This should probably be put in a subfolder with a better name (examples/connection?)

[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply go/.gitignore:5]: Oh I see
No need to rename, I think we will have only one example per client
Maybe: update the comment like `example binary`

File: go/src/lib.rs [hunk 9]
```diff
@@ -1,9 +1,434 @@
...
+
+#[repr(C)]
+pub struct RedisErrorFFI {
+    message: *const c_char,
+    error_type: ErrorType,
+}
+
+#[repr(u32)]
+pub enum ErrorType {
+    ClosingError = 0,
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:42]: Those are protobuf errors, on rust side there are much more error types. See `fn retry_method` in `types.rs` in `redis-rs`
I'm working on it in C#.

File: go/go.sum
[the remark this answers is not in the document]
Comment (aaron-congo) [reply go/go.sum:1]: This file should stay in the repo, I believe it is needed for users to resolve dependencies correctly. Other go repos include it, eg [go-redis](https://github.com/redis/go-redis/blob/master/go.sum). It is automatically updated when you add dependencies with `go get` and shouldn't ever have to be modified by hand.

File: go/glide/lib.h [hunk 10]
```diff
@@ -0,0 +1,144 @@
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+enum ErrorType {
```
[the remark this answers is not in the document]
Comment (aaron-congo) [reply go/glide/lib.h:6]: @Yury-Fridlyand why do you suggest the underscore? With cgo, enums are accessed by prefixing the C symbol with "enum_" already, so this would be accessed via `C.enum_ErrorType` in Go and it will not conflict with a Go type called ErrorType if we add it. If we have an extra underscore here then there would be an extra underscore to access it from Go as well, eg `C.enum__ErrorType`

File: go/src/lib.rs [hunk 11]
```diff
@@ -1,9 +1,434 @@
...
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let request_bytes = unsafe { std::slice::from_raw_parts(connection_request, request_len) };
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => {
+            let message_cstring = CString::new(err.to_string()).unwrap();
+            ConnectionResponse {
+                conn_ptr: std::ptr::null(),
+                error: &RedisErrorFFI {
```
Comment (aaron-congo) [inline go/src/lib.rs:91]: I see that we are allocating the ConnectionResponse struct and the GlideClient on the heap but I think we also need to allocate RedisErrorFFI and the RedisErrorFFI message on the heap.

File: go/src/lib.rs [hunk 12]
```diff
@@ -1,9 +1,434 @@
...
+    failure_callback: FailureCallback,
+) -> RedisResult<Client> {
+    let request =
+        connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes).unwrap();
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_name("GLIDE for Redis Go thread")
+        .build()?;
+    let _runtime_handle = runtime.enter();
+    let client = runtime.block_on(GlideClient::new(request)).unwrap(); // TODO - handle errors.
```
Comment (aaron-congo) [inline go/src/lib.rs:68]: Will the error handling be done in a separate PR, or is it straightforward to add to this PR? If we're doing it in another PR we will need to add a ticket for that

File: go/glide/glide.go [hunk 13]
```diff
@@ -1 +1,239 @@
...
+    resultChannelPtr := uintptr(unsafe.Pointer(&resultChannel))
+
+    args := []string{key}
+    cArgs := stringsToCStrings(args)
+    defer freeCStrings(cArgs)
+    requestType := C.uint32_t(GetString)
+    C.command(glideRedisClient.coreClient, C.uintptr_t(resultChannelPtr), requestType, C.uintptr_t(1), &cArgs[0])
+    resultPayload := <-resultChannel
+
+    return resultPayload.value, nil
```
[the remark this answers is not in the document]
Comment (jonathanl-bq) [reply go/glide/glide.go]: Will deal with this later when we add command logic

File: go/glide/lib.h [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  const char *message;
+  ErrorType error_type;
+} RedisErrorFFI;
+
+typedef struct ConnectionResponse {
+  const void *conn_ptr;
+  const struct RedisErrorFFI *error;
+} ConnectionResponse;
+
+typedef void (*SuccessCallback)(uintptr_t channel_address, const char *message);
```
[the remark this answers is not in the document]
Comment (jonathanl-bq) [reply go/glide/lib.h:28]: I'll do this when implementing command logic. I'll try to follow the design doc, but if it's not there, I'll just update it with this design.

File: go/glide/lib.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply go/glide/lib.h:6]: A C compiler won't compile such header I suppose due to naming conflict.
See example
https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/ne-processthreadsapi-process_information_class

File: go/go.sum
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply go/go.sum:1]: My mistake

File: go/glide/lib.h [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * Success callback that is called when a Redis command succeeds.
+ */
+typedef void (*SuccessCallback)(uintptr_t channel_address, const char *message);
+
+/**
+ * Failure callback that is called when a Redis command fails.
+ */
+typedef void (*FailureCallback)(uintptr_t channel_address, const struct RedisErrorFFI *error);
```
Comment (Yury-Fridlyand) [inline go/glide/lib.h:33]: To ease marshalling it is better to have
```suggestion
typedef void (*FailureCallback)(uintptr_t channel_address, ErrorType error_type, const char *message);
```
Otherwise we have to deal with memory allocations/deallocations for that pointer.

Comment (jonathanl-bq) [reply go/glide/lib.h:43]: I've already implemented the proper functions for freeing this struct and the string within. We can do this, but it goes against what's in the design doc and we'll have to make sure Aaron's work is aligned with this API change as well. I think we should discuss this first with the rest of the team before changing it.

File: go/src/lib.rs (hunk 12 above)
Comment (jonathanl-bq) [reply go/src/lib.rs]: I added it to this PR now.

File: go/glide/lib.h (hunk 15 above)
Comment (Yury-Fridlyand) [reply go/glide/lib.h:43]: It is ok to adjust the design noc now

Comment (jonathanl-bq) [reply go/glide/lib.h:43]: Is it? @aaron-congo are you aware of these changes?

File: go/glide/lib.h [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Creates a new client to the given address. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
+ *
+ * # Safety
+ *
+ * * `connection_request_bytes` must point to `connection_request_len` consecutive properly initialized bytes.
+ * * `connection_request_len` must not be greater than `isize::MAX`. See the safety documentation of [`std::slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html).
+ */
+const struct ConnectionResponse *create_client(const uint8_t *connection_request_bytes,
```
Comment (Yury-Fridlyand) [inline go/glide/lib.h:53]: Please, add a doc describing that `ConnectionResponse` should be freed afterwards regardless of the error

File: go/glide/lib.h [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+  const struct RedisErrorFFI *error;
+} ConnectionResponse;
+
+/**
+ * Success callback that is called when a Redis command succeeds.
+ */
+typedef void (*SuccessCallback)(uintptr_t channel_address, const char *message);
+
+/**
+ * Failure callback that is called when a Redis command fails.
```
Comment (Yury-Fridlyand) [inline go/glide/lib.h:41]: Please, add a doc describing that `RedisErrorFFI` should be freed afterwards

File: go/glide/lib.h (hunk 16 above)
Comment (jonathanl-bq) [reply go/glide/lib.h:53]: Added.

File: go/glide/lib.h (hunk 17 above)
Comment (jonathanl-bq) [reply go/glide/lib.h:41]: Added.

File: go/glide/glide.go [hunk 18]
```diff
@@ -1 +1,82 @@
+/**
+ * Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
+ */
 package glide
+
+/*
```
Comment (Yury-Fridlyand) [inline go/glide/glide.go:6]: Can we remove that comment?

File: go/glide/glide.go [hunk 19]
```diff
@@ -1 +1,82 @@
...
+func successCallback(channelPtr C.uintptr_t, message *C.char) {
+	goMessage := C.GoString(message)
+	goChannelPointer := uintptr(channelPtr)
+	resultChannel := *(*chan payload)(unsafe.Pointer(goChannelPointer))
+	resultChannel <- payload{value: goMessage, errMessage: nil}
+}
+
+//export failureCallback
+func failureCallback(channelPtr C.uintptr_t, errMessage *C.RedisErrorFFI) {
+	goMessage := C.GoString(errMessage.message)
```
Comment (Yury-Fridlyand) [inline go/glide/glide.go:53]: You are loosing error type there. Any plans to handle it?

File: go/glide/glide.go [hunk 20]
```diff
@@ -1 +1,82 @@
...
+		defer C.free_error(response.error)
+		return fmt.Errorf(C.GoString(response.error.message))
+	}
+	glideRedisClient.coreClient = response.conn_ptr
+	return nil
+}
+
+func (glideRedisClient *GlideRedisClient) CloseClient() error {
+	if glideRedisClient.coreClient == nil {
+		return fmt.Errorf("Cannot close glide client before it has been created.")
```
Comment (Yury-Fridlyand) [inline go/glide/glide.go:78]: ... or already closed

File: go/glide/glide.go [hunk 21]
```diff
@@ -1 +1,82 @@
...
+	}
+	C.close_client(glideRedisClient.coreClient)
```
Comment (Yury-Fridlyand) [inline go/glide/glide.go:80]:
```suggestion
	C.close_client(glideRedisClient.coreClient)
	glideRedisClient.coreClient = nil
```

File: go/glide/glide.go [hunk 22]
```diff
@@ -1 +1,82 @@
...
+	}
+	byteCount := len(marshalledRequest)
+	requestBytes := C.CBytes(marshalledRequest)
+	response := (*C.struct_ConnectionResponse)(C.create_client((*C.uchar)(requestBytes), C.uintptr_t(byteCount), (C.SuccessCallback)(unsafe.Pointer(C.successCallback)), (C.FailureCallback)(unsafe.Pointer(C.failureCallback))))
+	defer C.free_connection_response(response)
+	if response.error != nil {
+		defer C.free_error(response.error)
+		return fmt.Errorf(C.GoString(response.error.message))
+	}
+	glideRedisClient.coreClient = response.conn_ptr
```
Comment (Yury-Fridlyand) [inline go/glide/glide.go:72]: Why here `=` but not `:=`? A go tricks?

File: go/glide/lib.h (hunk 17 above)
Comment (Yury-Fridlyand) [reply go/glide/lib.h:41]: Did you update this file after that? I see it on rs part, but it is not reflected there.

File: go/src/lib.rs [hunk 23]
```diff
@@ -1,9 +1,172 @@
...
+
+/// A Redis error.
+#[repr(C)]
+pub struct RedisErrorFFI {
+    message: *const c_char,
+    error_type: ErrorType,
+}
+
+/// FFI compatible version of the ErrorType enum defined in protobuf.
+// TODO: Update when command errors are implemented
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:41]: Add https://github.com/aws/glide-for-redis/pull/1077 reference for TODO
We will remove this enum after #1077

File: go/src/lib.rs [hunk 24]
```diff
@@ -1,9 +1,172 @@
...
+    connection_request_len: usize,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let request_bytes =
+        unsafe { std::slice::from_raw_parts(connection_request_bytes, connection_request_len) };
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            error: Box::into_raw(Box::new(err)) as *const RedisErrorFFI,
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:120]: Should you wrap this line with unsafe?
```suggestion
            error: unsafe { Box::into_raw(Box::new(err)) as *const RedisErrorFFI },
```
If yes, line 123 too

File: go/src/lib.rs [hunk 25]
```diff
@@ -1,9 +1,172 @@
...
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, RedisErrorFFI> {
+    let request = connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes)
+        .map_err(|err| {
+            let msg = CString::new(err.to_string()).unwrap();
+            RedisErrorFFI {
+                message: msg.into_raw(),
+                error_type: ErrorType::ConnectionError,
+            }
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:71]: Do you want to make a function for this? You repeat it 3 times

File: go/src/lib.rs [hunk 26]
```diff
@@ -1,9 +1,172 @@
...
+pub enum ErrorType {
+    ClosingError = 0,
+    RequestError = 1,
+    TimeoutError = 2,
+    ExecAbortError = 3,
+    ConnectionError = 4,
+}
+
+/// The glide client.
+#[allow(dead_code)]
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:52]: why?

File: go/src/lib.rs [hunk 27]
```diff
@@ -1,9 +1,172 @@
...
+    })?;
+    Ok(Client {
+        client,
+        success_callback,
+        failure_callback,
+        runtime,
+    })
+}
+
+/// Creates a new client to the given address. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:100]:
```suggestion
/// Creates a new client with the given configuration. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
```

File: go/src/lib.rs (hunk 24 above)
Comment (jonathanl-bq) [reply go/src/lib.rs:120]: In general, it's safe to create raw pointers in Rust. Dereferencing is unsafe, however.

File: go/src/lib.rs (hunk 26 above)
Comment (jonathanl-bq) [reply go/src/lib.rs:52]: To pass CI checks. We'll remove it later.

File: go/glide/glide.go (hunk 22 above)
Comment (jonathanl-bq) [reply go/glide/glide.go:72]: := is used when assigning to a new variable

File: go/glide/glide.go (hunk 18 above)
Comment (jonathanl-bq) [reply go/glide/glide.go:6]: Nope. It's actually read by cgo and is required to work with it.

File: go/glide/lib.h (hunk 15 above)
Comment (aaron-congo) [reply go/glide/lib.h:43]: I wasn't aware. We need the RedisErrorFFI and a mechanism to free it for the connection request, since we are not using a callback and C functions can only return a single result. If we need these things anyways, should we just use it in the failure callback as planned?

File: go/examples/main.go (hunk 6 above)
Comment (aaron-congo) [reply go/examples/main.go:15]: I have these changes in my connect branch and can change the go files to integrate with that branch in my upcoming connect PR. This PR should be focused more on the Rust code than the Go code since the Go code will change when I integrate

File: go/glide/lib.h [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+ */
+typedef struct RedisErrorFFI {
+  const char *message;
+  enum ErrorType error_type;
+} RedisErrorFFI;
+
+/**
+ * The connection response.
+ *
+ * It contains either a connection or an error. It is represented as a struct instead of an enum for ease of use in the wrapper language.
```
Comment (aaron-congo) [inline go/glide/lib.h:28]: Do you mean instead of a union?
```suggestion
 * It contains either a connection or an error. It is represented as a struct instead of a union for ease of use in the wrapper language.
```

File: go/glide/lib.h [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+ */
+typedef void (*SuccessCallback)(uintptr_t channel_address, const char *message);
+
+/**
+ * Failure callback that is called when a Redis command fails.
+ */
+typedef void (*FailureCallback)(uintptr_t channel_address, const struct RedisErrorFFI *error);
+
+/**
+ * Creates a new client to the given address. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
```
Comment (aaron-congo) [inline go/glide/lib.h:46]: What is the given string you are referring to here? Would it make sense to move the success callback sentence here to be above the SuccessCallback doc instead of the create_client doc?

File: go/glide/lib.h [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+ * * `client_ptr` must not be null.
+ */
+void close_client(const void *client_ptr);
+
+/**
+ * Deallocates a `ConnectionResponse`.
+ *
+ * # Safety
+ *
+ * * `connection_response_ptr` must be able to be safely casted to a valid `Box<RedisErrorFFI>` via `Box::from_raw`. See the safety documentation of [`std::boxed::Box::from_raw`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw).
```
Comment (aaron-congo) [inline go/glide/lib.h:73]:
```suggestion
 * * `connection_response_ptr` must be able to be safely casted to a valid `Box<ConnectionResponse>` via `Box::from_raw`. See the safety documentation of [`std::boxed::Box::from_raw`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw).
```

File: go/go.mod [hunk 31]
```diff
@@ -10,6 +10,7 @@ require (
 
 require (
 	github.com/BurntSushi/toml v1.2.1 // indirect
+	github.com/golang/protobuf v1.5.3 // indirect
```
Comment (aaron-congo) [inline go/go.mod:13]: Do you know why this is added? On my connect branch it is not needed since protobuf is already listed on line 7. If you do `go mod tidy` does this get removed?

File: go/src/lib.rs [hunk 32]
```diff
@@ -1,9 +1,172 @@
...
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, RedisErrorFFI> {
+    let request = connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes)
+        .map_err(|err| {
+            let msg = CString::new(err.to_string()).unwrap();
+            RedisErrorFFI {
+                message: msg.into_raw(),
+                error_type: ErrorType::ConnectionError,
```
Comment (aaron-congo) [inline go/src/lib.rs:70]: The Java code has a test that connects to an invalid address and then asserts that a ClosingException occurred, should this be ClosingError instead of ConnectionError?
https://github.com/Bit-Quill/glide-for-redis/blob/main/java/integTest/src/test/java/glide/ErrorHandlingTests.java#L34

File: go/src/lib.rs [hunk 33]
```diff
@@ -1,9 +1,172 @@
...
+    let client_ptr = unsafe { Box::from_raw(client_ptr as *mut Client) };
+    let _runtime_handle = client_ptr.runtime.enter();
+    drop(client_ptr);
+}
+
+/// Deallocates a `ConnectionResponse`.
+///
+/// # Safety
+///
+/// * `connection_response_ptr` must be able to be safely casted to a valid `Box<RedisErrorFFI>` via `Box::from_raw`. See the safety documentation of [`std::boxed::Box::from_raw`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw).
```
Comment (aaron-congo) [inline go/src/lib.rs:147]:
```suggestion
/// * `connection_response_ptr` must be able to be safely casted to a valid `Box<ConnectionResponse>` via `Box::from_raw`. See the safety documentation of [`std::boxed::Box::from_raw`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw).
```

File: go/glide/lib.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (aaron-congo) [reply go/glide/lib.h]: What would the naming conflict be between? The ErrorType symbol here and what else? As mentioned, I don't think it will conflict with any Go symbols

File: go/src/lib.rs (hunk 32 above)
Comment (Yury-Fridlyand) [reply go/src/lib.rs:70]: @shachlanAmazon said no in https://github.com/aws/glide-for-redis/pull/1077#discussion_r1514297208, bit I hope he will change his mind later.
We can keep it as `ConnectionError` and change later

File: go/src/lib.rs (hunk 23 above)
Comment (Yury-Fridlyand) [reply go/src/lib.rs:41]: 1077 merged, you can update your code and remove TODO

File: go/glide/lib.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply go/glide/lib.h]: `ErrorType` is enum and typedef
Well, if this file is autogenerated, we have to keep this as is.

File: go/glide/lib.h (hunk 29 above)
Comment (jonathanl-bq) [reply go/glide/lib.h:43]: I'm trying to remember where I copied this doc comment from originally. I think it was the C# implementation? I can still move it if you like.

File: go/src/lib.rs [hunk 34]
```diff
@@ -1,9 +1,160 @@
...
+/// `error` should be manually freed by calling `free_error` after this callback is invoked, otherwise a memory leak will occur.
+pub type FailureCallback = unsafe extern "C" fn(
+    channel_address: usize,
+    error_message: *const c_char,
+    error_type: RequestErrorType,
+) -> ();
+
+/// The connection response.
+///
+/// It contains either a connection or an error. It is represented as a struct instead of an enum for ease of use in the wrapper language.
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:32]: Maybe warn here too about freeing this struct

File: go/src/lib.rs [hunk 35]
```diff
@@ -1,9 +1,160 @@
...
+///
+/// * `connection_response_ptr` must be able to be safely casted to a valid `Box<ConnectionResponse>` via `Box::from_raw`. See the safety documentation of [`std::boxed::Box::from_raw`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw).
+/// * `connection_response_ptr` must not be null.
+#[no_mangle]
+pub unsafe extern "C" fn free_connection_response(
+    connection_response_ptr: *const ConnectionResponse,
+) {
+    let connection_response =
+        unsafe { Box::from_raw(connection_response_ptr as *mut ConnectionResponse) };
+    drop(connection_response);
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:147]: Don't you need to release `connection_response.error_message` here? I think you can call `free_error`

File: go/src/lib.rs [hunk 36]
```diff
@@ -1,9 +1,160 @@
...
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            error_message: CString::into_raw(CString::new(err.message).unwrap()),
+            error_type: err.error_type,
+        },
+        Ok(client) => ConnectionResponse {
+            conn_ptr: Box::into_raw(Box::new(client)) as *const c_void,
+            error_message: std::ptr::null(),
+            error_type: RequestErrorType::Unspecified,
```
Comment (Yury-Fridlyand) [inline go/src/lib.rs:116]: Should we ash Shachar to define `NoError` enum variant? (later)

**Review by Yury-Fridlyand — APPROVED** [review]

**Comment by Yury-Fridlyand** [comment]
CI is red

File: go/src/lib.rs (hunk 36 above)
Comment (jonathanl-bq) [reply go/src/lib.rs:116]: I think it would be nice. We'll see what he thinks though.

File: go/src/lib.rs (hunk 35 above)
Comment (jonathanl-bq) [reply go/src/lib.rs:147]: We need to be able to free the connection response separately from the error contained within it. When we get back a connection pointer, the error is null, so when we try to free the null pointer, we get a seg fault.

File: go/src/lib.rs (hunk 34 above)
Comment (jonathanl-bq) [reply go/src/lib.rs:32]: Good idea.

File: go/src/lib.rs (hunk 35 above)
Comment (Yury-Fridlyand) [reply go/src/lib.rs:147]: Can you add ?
```rs
if connection_response.error_message != std::ptr::null()
    free_error(connection_response.error_message);
```

Comment (jonathanl-bq) [reply go/src/lib.rs:151]: Okay. I'll do this and also make free_error private instead of exposing it to the wrapper language.

File: go/src/lib.rs [hunk 37]
```diff
@@ -1,9 +1,168 @@
...
+#[no_mangle]
+pub unsafe extern "C" fn close_client(client_ptr: *const c_void) {
+    let client_ptr = unsafe { Box::from_raw(client_ptr as *mut Client) };
+    let _runtime_handle = client_ptr.runtime.enter();
+    drop(client_ptr);
+}
+
+/// Deallocates a `ConnectionResponse`.
+///
+/// This function does not free the contained error, which needs to be freed separately using `free_error` afterwards to avoid memory leaks.
```
Comment (aaron-congo) [inline go/src/lib.rs:139]: Looks like this doc will have to get updated since it now frees the error

File: go/glide/lib.h [hunk 38]
```diff
@@ -78,3 +78,13 @@ void close_client(const void *client_ptr);
...
+
+/**
+ * Deallocates an error message `CString`.
+ *
+ * # Safety
+ *
+ * * `error_msg_ptr` must be able to be safely casted to a valid `CString` via `CString::from_raw`. See the safety documentation of [`std::ffi::CString::from_raw`](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_raw).
+ * * `error_msg_ptr` must not be null.
+ */
+void free_error(const char *error_msg_ptr);
```
Comment (Yury-Fridlyand) [inline go/glide/lib.h:90]: Why re-expose?

Comment (jonathanl-bq) [reply go/glide/lib.h:90]: I assumed that the failure callback would use it for the error it gets for commands. Not relevant for connection logic though, so I guess it can be reverted. Up to you, since you have all the changes in your branch now.

**Comment by jonathanl-bq** [comment]
Closing as all relevant changes are integrated into other branches now.
